dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -142,147 +131,155 @@ bool DMFileReader::getSkippedRows(size_t & skip_rows)
 }
 
+// Skip the block which should be returned by next read()
 size_t DMFileReader::skipNextBlock()
 {
-    // Go to next available pack.
-    size_t skip;
-    if (!getSkippedRows(skip))
+    // find the first pack which is used
+    size_t skip_rows;
+    if (!getSkippedRows(skip_rows))
         return 0;
 
-    // Find the next contiguous packs will be read in next read,
-    // let next_pack_id point to the next pack of the contiguous packs.
-    // For example, if we have 10 packs, use_packs is [0, 1, 1, 0, 1, 1, 0, 0, 1, 1],
-    // and now next_pack_id is 1, then we will skip 2 packs(index 1 and 2), and next_pack_id will be 3.
-    const size_t read_pack_limit = read_one_pack_every_time ? 1 : 0;
-    const std::vector<RSResult> & handle_res = pack_filter.getHandleRes();
-    const RSResult expected_handle_res = handle_res[next_pack_id];
-    auto & use_packs = pack_filter.getUsePacks();
+    // move forward next_pack_id and next_row_offset
+    size_t read_rows = getReadRows();
+    if (read_rows == 0)
+        return 0;
+
+    addSkippedRows(read_rows);
+    // When we read dmfile, if the previous pack is not read,
+    // then we should seek to the right offset of dmfile.
+    std::fill(must_seek.begin(), must_seek.end(), true);
+
+    scan_context->late_materialization_skip_rows += read_rows;
+    return read_rows;
+}
+
+// Get the number of rows to read in the next block
+// Move forward next_pack_id and next_row_offset
+size_t DMFileReader::getReadRows()
+{
+    const auto & use_packs = pack_filter.getUsePacksConst();
     size_t start_pack_id = next_pack_id;
+    // When read_one_pack_every_time is true, we can just read one pack every time.
+    // std::numeric_limits<size_t>::max() means no limit
+    size_t read_pack_limit = read_one_pack_every_time ? 1 : std::numeric_limits<size_t>::max();
     const auto & pack_stats = dmfile->getPackStats();
     size_t read_rows = 0;
     for (; next_pack_id < use_packs.size() && use_packs[next_pack_id] && read_rows < rows_threshold_per_read;
          ++next_pack_id)
     {
-        if (read_pack_limit != 0 && next_pack_id - start_pack_id >= read_pack_limit)
-            break;
-        if (enable_handle_clean_read && handle_res[next_pack_id] != expected_handle_res)
+        if (next_pack_id - start_pack_id >= read_pack_limit)
             break;
-
         read_rows += pack_stats[next_pack_id].rows;
     }
 
-    addSkippedRows(read_rows);
     next_row_offset += read_rows;
-
-    // When we read dmfile, if the previous pack is not read,
-    // then we should seek to the right offset of dmfile.
-    // So if skip some packs successfully,
-    // then we set the last pack to false to indicate that we should seek before read.
-    if (likely(read_rows > 0))
-        use_packs[next_pack_id - 1] = false;
-
-    scan_context->late_materialization_skip_rows += read_rows;
     return read_rows;
 }
 
 Block DMFileReader::readWithFilter(const IColumn::Filter & filter)
```

Comment:
Do we have unit tests for this class after this refactor?

Reply:
dbms/src/Storages/DeltaMerge/tests/gtest_skippable_block_input_stream.cpp

There are unit tests covered, and I have refine the tests.